### PR TITLE
4dmm

### PR DESCRIPTION
**Pull request by dding3**
This PR is split from https://github.com/intel-analytics/analytics-zoo/pull/966/files and used to 
* Support 4d tensors mm
* Support math operations with different dims
* Support math operations for singleton dim

### CODE REVIEW

**Comment by dding3** [comment]
jekins passed. http://172.168.2.101:8080/view/ZOO-PR/job/ZOO-PR-Validation/1898/

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/autograd/math.scala [hunk 1]
```diff
@@ -288,11 +297,13 @@ object AutoGrad {
...
     }
 
-
     val mm = InternalMM[T](transA = transposeX,
       transB = transposeY)
     val kmm = new KerasLayerWrapper[T](mm.asInstanceOf[AbstractModule[Activity, Activity, T]])
-    kmm.from(x, y)
+
+    if (xShape.length > 3 || yShape.length > 3) {
+      TimeDistributed(kmm.asInstanceOf[KerasLayer[Activity, Tensor[T], T]]).from(xx, yy)
```
Comment (zhichao-li) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/autograd/math.scala]: should we update the logic of `axes` accordingly or add some doc for the special use case for this 4dmm?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/autograd/math.scala [hunk 2]
```diff
@@ -565,6 +574,13 @@ class Variable[T: ClassTag] private[zoo] (private[zoo] var node: ModuleNode[T],
     Variable(o.inputs(this.node))
   }
 
+  def expand(sizes: List[Int]): Variable[T] = {
```
Comment (zhichao-li) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/autograd/math.scala]: add some doc for this op? and what's the difference between `expand` and `expandDims`?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/InternalMM.scala [hunk 3]
```diff
@@ -98,6 +112,22 @@ private[zoo] class InternalMM[T: ClassTag](
...
     var (ma, mb) = checkInputFormat(input)
 
+    if (ma.dim() > 2 && ma.size(1) != mb.size(1)) {
+      require(mb.dim() == 3, "second input tensor must be 3D" +
+        s"second input dim ${mb.dim()}")
+      // with different batch dim
+      if (ma.size(1) == 1) {
+        expandLayer = InternalExpand(mb.size())
+        ma = expandLayer.forward(ma)
+      } else if (mb.size(1) == 1) {
```
Comment (zhichao-li) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/InternalMM.scala]: should we extract these as a method as it has been duplicated twice within the same file?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/InternalMM.scala [hunk 4]
```diff
@@ -85,6 +86,19 @@ private[zoo] class InternalMM[T: ClassTag](
       require(ma.size(3) == mb.size(2), "matrix sizes do not match" +
         s"the matrix sizes are ${ma.size(3)} and ${mb.size(2)}")
 
+      // with different batch dim
+      if (ma.size(1) != mb.size(1)) {
+        if (ma.size(1) == 1) {
```
Comment (zhichao-li) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/InternalMM.scala]: Maybe we should add some doc for this `InternalMM` that difference batchSize can be broadcast and supported.

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/InternalTimeDistributed.scala [hunk 5]
```diff
@@ -46,11 +46,11 @@ import scala.reflect.ClassTag
  */
 
 private[zoo] class InternalTimeDistributed[T: ClassTag](
-    val layer: AbstractModule[Tensor[T], Tensor[T], T],
+    val layer: AbstractModule[Activity, Tensor[T], T],
```
Comment (zhichao-li) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/InternalTimeDistributed.scala]: Add some doc for this? i.e no nested Table is supported?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/InternalTimeDistributed.scala [hunk 6]
```diff
@@ -128,59 +150,60 @@ private[zoo] class InternalTimeDistributed[T: ClassTag](
...
+    layer.accGradParameters(input, gradOutput)
+    resizeActivity(input, oriSizes)
+    gradOutput.resize(_gradOutputSize)
+  }
 
+  override def backward(input: Activity, gradOutput: Tensor[T]): Activity = {
+    val st = System.nanoTime
+    updateGradInput(input, gradOutput)
+    accGradParameters(input, gradOutput)
+    backwardTime += System.nanoTime - st
```
Comment (zhichao-li) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/InternalTimeDistributed.scala]: what's the purpose for calc time here?

**Comment by zhichao-li** [comment]
It's fine to enrich this API in this way for now, but maybe we should add more document to address the special usage which is not that general as expected.

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/InternalTimeDistributed.scala (hunk 6 above)
Comment (dding3) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/InternalTimeDistributed.scala]: It's copied from previous implementation. I think it's used for internal performance testing.

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/autograd/math.scala (hunk 1 above)
Comment (dding3) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/autograd/math.scala]: Updated

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/autograd/math.scala (hunk 2 above)
Comment (dding3) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/autograd/math.scala]: Added

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/InternalTimeDistributed.scala (hunk 5 above)
Comment (dding3) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/InternalTimeDistributed.scala]: Added

**Comment by zhichao-li** [comment]
LGTM

**Comment by dding3** [comment]
jekins passed. http://172.168.2.101:8080/view/ZOO-PR/job/ZOO-PR-Validation/1926/

Thanks for your comments.

**Pull request merged**